crates/rlbot/src/agents.rs

```diff
@@ -36,160 +28,209 @@ pub enum AgentError {
     PacketParseError(#[from] crate::RLBotError),
 }
 
-/// A queue of packets to be sent to RLBotServer
-pub struct PacketQueue {
-    internal_queue: Vec<Packet>,
-}
-
-impl PacketQueue {
-    pub fn new() -> Self {
-        PacketQueue {
-            internal_queue: Vec::with_capacity(16),
-        }
-    }
-    pub fn push(&mut self, packet: impl Into<Packet>) {
-        self.internal_queue.push(packet.into());
-    }
-    fn empty(&mut self) -> Vec<Packet> {
-        mem::take(&mut self.internal_queue)
-    }
-}
-
-/// Run multiple agents on one thread each. They share a connection.
+/// Run multiple agents with n agents per thread. They share a connection.
 /// Ok(()) means a successful exit; one of the bots received a None packet.
+///
+/// # Errors
+///
+/// Returns an error if an agent panics or if there is an error with the connection.
+///
+/// # Panics
+///
+/// Panics if a thread can't be spawned for each agent.
 pub fn run_agents<T: Agent>(
-    connection_settings: ConnectionSettings,
+    agent_id: String,
+    wants_ball_predictions: bool,
+    wants_comms: bool,
     mut connection: RLBotConnection,
 ) -> Result<(), AgentError> {
-    connection.send_packet(connection_settings)?;
+    connection.send_packet(ConnectionSettings {
+        agent_id,
+        wants_ball_predictions,
+        wants_comms,
+        close_between_matches: true,
+    })?;
+
+    let StartingInfo {
+        controllable_team_info,
+        match_configuration,
+        field_info,
+    } = connection.get_starting_info()?;
 
-    let mut packets_to_process = VecDeque::new();
+    if controllable_team_info.controllables.is_empty() {
+        // run no bots? no problem, done
+        return Ok(());
+    }
 
-    // Wait for ControllableTeamInfo to know which indices we control
-    let controllable_team_info = loop {
-        let packet = connection.recv_packet()?;
-        if let Packet::ControllableTeamInfo(x) = packet {
-            break x;
-        } else {
-            packets_to_process.push_back(packet);
-            continue;
-        }
-    };
+    let match_config = Arc::new(match_configuration);
+    let field_info = Arc::new(field_info);
```

Comment:
field_info is overwritten but match_configuration isn't

Reply:
i just like `match_config` because it's shorter. `match_configuration`, or `match_config`? I can change everything to one or the other.

Reply:
do whatever matches the struct generated from the spec :)